### ISSUE TITLE
Cross-linking parameters fails when using `@overload`

### ISSUE THREAD

**Issue by mhostetter** [jbms/sphinx-immaterial#85]
Thanks for the fix for #82! When testing the latest changes, I noticed my build was failing. I narrowed down the issue to be related to parameter cross-referencing with overloaded functions. I created a minimal reproducer below.

### `index.rst`
```rst
Test
====

.. autofunction:: my_module.my_function

.. autofunction:: my_module.my_overloaded_function
```

### `my_module.py`

```python
from typing import overload


def my_function(x: int, y: int) -> int:
    """
    Adds two ints.
    """
    return x + y


@overload
def my_overloaded_function(x: int, y: int) -> int:
    ...
@overload
def my_overloaded_function(x: str, y: str) -> str:
    ...
def my_overloaded_function(x, y):
    """
    Overloaded function that adds two ints or two strings.
    """
    return x + y
```

### Build output

```
Running Sphinx v4.5.0
loading translations [en]... locale_dir /mnt/c/Users/matth/repos/galois/docs/locales/en/LC_MESSAGES does not exists
locale_dir /mnt/c/Users/matth/repos/galois/docs/locales/en/LC_MESSAGES does not exists
done
making output directory... done
locale_dir /mnt/c/Users/matth/repos/galois/docs/locales/en/LC_MESSAGES does not exists
[autosummary] generating autosummary for: index.rst
loading intersphinx inventory from https://docs.python.org/3/objects.inv...
loading intersphinx inventory from https://numpy.org/doc/stable/objects.inv...
myst v0.17.2: MdParserConfig(commonmark_only=False, gfm_only=False, enable_extensions=[], linkify_fuzzy_links=True, dmath_allow_labels=True, dmath_allow_space=True, dmath_allow_digits=True, dmath_double_inline=False, update_mathjax=True, mathjax_classes='tex2jax_process|mathjax_process|math|output_area', disable_syntax=[], all_links_external=False, url_schemes=('http', 'https', 'mailto', 'ftp'), ref_domains=None, highlight_code_blocks=True, number_code_blocks=[], title_to_header=False, heading_anchors=None, heading_slug_func=None, footnote_transition=True, sub_delimiters=('{', '}'), words_per_minute=200)
locale_dir /mnt/c/Users/matth/repos/galois/docs/locales/en/LC_MESSAGES does not exists
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 1 source files that are out of date
updating environment: locale_dir /mnt/c/Users/matth/repos/galois/docs/locales/en/LC_MESSAGES does not exists
[new config] 1 added, 0 changed, 0 removed
reading sources... [100%] index

Traceback (most recent call last):
  File "/home/matt/.local/lib/python3.8/site-packages/sphinx/cmd/build.py", line 276, in build_main
    app.build(args.force_all, filenames)
  File "/home/matt/.local/lib/python3.8/site-packages/sphinx/application.py", line 330, in build
    self.builder.build_update()
  File "/home/matt/.local/lib/python3.8/site-packages/sphinx/builders/__init__.py", line 286, in build_update
    self.build(to_build,
  File "/home/matt/.local/lib/python3.8/site-packages/sphinx/builders/__init__.py", line 300, in build
    updated_docnames = set(self.read())
  File "/home/matt/.local/lib/python3.8/site-packages/sphinx/builders/__init__.py", line 407, in read
    self._read_serial(docnames)
  File "/home/matt/.local/lib/python3.8/site-packages/sphinx/builders/__init__.py", line 428, in _read_serial
    self.read_doc(docname)
  File "/home/matt/.local/lib/python3.8/site-packages/sphinx/builders/__init__.py", line 468, in read_doc
    doctree = read_doc(self.app, self.env, self.env.doc2path(docname))
  File "/home/matt/.local/lib/python3.8/site-packages/sphinx/io.py", line 181, in read_doc
    pub.publish()
  File "/usr/lib/python3/dist-packages/docutils/core.py", line 217, in publish
    self.document = self.reader.read(self.source, self.parser,
  File "/home/matt/.local/lib/python3.8/site-packages/sphinx/io.py", line 101, in read
    self.parse()
  File "/usr/lib/python3/dist-packages/docutils/readers/__init__.py", line 77, in parse
    self.parser.parse(self.input, document)
  File "/home/matt/.local/lib/python3.8/site-packages/sphinx/parsers.py", line 89, in parse
    self.statemachine.run(inputlines, document, inliner=self.inliner)
  File "/usr/lib/python3/dist-packages/docutils/parsers/rst/states.py", line 171, in run
    results = StateMachineWS.run(self, input_lines, input_offset,
  File "/usr/lib/python3/dist-packages/docutils/statemachine.py", line 241, in run
    context, next_state, result = self.check_line(
  File "/usr/lib/python3/dist-packages/docutils/statemachine.py", line 459, in check_line
    return method(match, context, next_state)
  File "/usr/lib/python3/dist-packages/docutils/parsers/rst/states.py", line 2770, in underline
    self.section(title, source, style, lineno - 1, messages)
  File "/usr/lib/python3/dist-packages/docutils/parsers/rst/states.py", line 328, in section
    self.new_subsection(title, lineno, messages)
  File "/usr/lib/python3/dist-packages/docutils/parsers/rst/states.py", line 394, in new_subsection
    newabsoffset = self.nested_parse(
  File "/usr/lib/python3/dist-packages/docutils/parsers/rst/states.py", line 282, in nested_parse
    state_machine.run(block, input_offset, memo=self.memo,
  File "/usr/lib/python3/dist-packages/docutils/parsers/rst/states.py", line 197, in run
    results = StateMachineWS.run(self, input_lines, input_offset)
  File "/usr/lib/python3/dist-packages/docutils/statemachine.py", line 241, in run
    context, next_state, result = self.check_line(
  File "/usr/lib/python3/dist-packages/docutils/statemachine.py", line 459, in check_line
    return method(match, context, next_state)
  File "/usr/lib/python3/dist-packages/docutils/parsers/rst/states.py", line 2345, in explicit_markup
    self.explicit_list(blank_finish)
  File "/usr/lib/python3/dist-packages/docutils/parsers/rst/states.py", line 2370, in explicit_list
    newline_offset, blank_finish = self.nested_list_parse(
  File "/usr/lib/python3/dist-packages/docutils/parsers/rst/states.py", line 319, in nested_list_parse
    state_machine.run(block, input_offset, memo=self.memo,
  File "/usr/lib/python3/dist-packages/docutils/parsers/rst/states.py", line 197, in run
    results = StateMachineWS.run(self, input_lines, input_offset)
  File "/usr/lib/python3/dist-packages/docutils/statemachine.py", line 241, in run
    context, next_state, result = self.check_line(
  File "/usr/lib/python3/dist-packages/docutils/statemachine.py", line 459, in check_line
    return method(match, context, next_state)
  File "/usr/lib/python3/dist-packages/docutils/parsers/rst/states.py", line 2648, in explicit_markup
    nodelist, blank_finish = self.explicit_construct(match)
  File "/usr/lib/python3/dist-packages/docutils/parsers/rst/states.py", line 2355, in explicit_construct
    return method(self, expmatch)
  File "/usr/lib/python3/dist-packages/docutils/parsers/rst/states.py", line 2097, in directive
    return self.run_directive(
  File "/usr/lib/python3/dist-packages/docutils/parsers/rst/states.py", line 2147, in run_directive
    result = directive_instance.run()
  File "/home/matt/.local/lib/python3.8/site-packages/sphinx/ext/autodoc/directive.py", line 165, in run
    result = parse_generated_content(self.state, params.result, documenter)
  File "/home/matt/.local/lib/python3.8/site-packages/sphinx/ext/autodoc/directive.py", line 112, in parse_generated_content
    state.nested_parse(content, 0, node)
  File "/usr/lib/python3/dist-packages/docutils/parsers/rst/states.py", line 282, in nested_parse
    state_machine.run(block, input_offset, memo=self.memo,
  File "/usr/lib/python3/dist-packages/docutils/parsers/rst/states.py", line 197, in run
    results = StateMachineWS.run(self, input_lines, input_offset)
  File "/usr/lib/python3/dist-packages/docutils/statemachine.py", line 241, in run
    context, next_state, result = self.check_line(
  File "/usr/lib/python3/dist-packages/docutils/statemachine.py", line 459, in check_line
    return method(match, context, next_state)
  File "/usr/lib/python3/dist-packages/docutils/parsers/rst/states.py", line 2343, in explicit_markup
    nodelist, blank_finish = self.explicit_construct(match)
  File "/usr/lib/python3/dist-packages/docutils/parsers/rst/states.py", line 2355, in explicit_construct
    return method(self, expmatch)
  File "/usr/lib/python3/dist-packages/docutils/parsers/rst/states.py", line 2097, in directive
    return self.run_directive(
  File "/usr/lib/python3/dist-packages/docutils/parsers/rst/states.py", line 2147, in run_directive
    result = directive_instance.run()
  File "/home/matt/.local/lib/python3.8/site-packages/sphinx/domains/__init__.py", line 281, in run
    return super().run()
  File "/home/matt/.local/lib/python3.8/site-packages/sphinx_immaterial/apidoc_formatting.py", line 178, in run
    nodes = orig_run(self)
  File "/home/matt/.local/lib/python3.8/site-packages/sphinx/directives/__init__.py", line 206, in run
    self.after_content()
  File "/home/matt/.local/lib/python3.8/site-packages/sphinx_immaterial/python_domain_fixes.py", line 636, in after_content
    _cross_link_parameters(
  File "/home/matt/.local/lib/python3.8/site-packages/sphinx_immaterial/python_domain_fixes.py", line 576, in _cross_link_parameters
    assert len(signodes) == len(symbols)
AssertionError

Exception occurred:
  File "/home/matt/.local/lib/python3.8/site-packages/sphinx_immaterial/python_domain_fixes.py", line 576, in _cross_link_parameters
    assert len(signodes) == len(symbols)
AssertionError
The full traceback has been saved in /tmp/sphinx-err-o662a6eq.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
```

**Comment by 2bndy5** [comment]
I also ran into a similar problem when using autodoc for a c-extension (wrapped using pybind11) that has overloaded functions exposed in python. So, my situation is a result of using pybind11 to add docstrings (including function signatures) from C++ code and autodoc's limited capability handling c-extensions' overloaded functions. I could describe more, but I think my problem is a separate issue since the `assert len(signodes) == len(symbols)` is satisfied (overloaded functions' docstrings are grouped together by autodoc).

It seems it could be useful to turn off the parameters' cross-linking feature since sphinx (& extensions) wasn't really designed to do that.

**Comment by jbms** [comment]
> I also ran into a similar problem when using autodoc for a c-extension (wrapped using pybind11) that has overloaded functions exposed in python. So, my situation is a result of using pybind11 to add docstrings (including function signatures) from C++ code and autodoc's limited capability handling c-extensions' overloaded functions. I could describe more, but I think my problem is a separate issue since the `assert len(signodes) == len(symbols)` is satisfied (overloaded functions' docstrings are grouped together by autodoc).
> 
> It seems it could be useful to turn off the parameters' cross-linking feature since sphinx (& extensions) wasn't really designed to do that.

That is something we definitely should support, but I agree that it is a separate issue.

This issue is about one `py:function` directive with multiple signatures (but a single body).

This parameter cross-linking feature is basically a port from a hackier implementation in the tensorstore docs.  (In the tensorstore docs it was also coupled with my custom autosummary extension --- I'm trying to decouple a lot of functionality for inclusion in this theme.)

Tensorstore also uses pybind11 to define overloaded functions, but in that case each overload is documented using a separate `py:function` directive since they each have their own documentation.

Normally the Sphinx Python domain just uses the normal symbol name (e.g. `modname.func`) as the identifier for the object, but that leads to a conflict if you want to document multiple overloads.  In tensorstore I solve this by requiring the user specify an "overload" identifier for each overload if there is more than one.  Then the full identifier of the object becomes `modname.func(overload_id)`.  The TOC will also show `func(overload_id)`.  Often I will choose the overload id to match the name of the first parameter, if that is a suitable unique identifier, since then the syntax makes more sense.

Ideally all of the tensorstore documentation functionality can be ported over to this theme.

**Comment by 2bndy5** [comment]
> Tensorstore also uses pybind11 to define overloaded functions, but in that case each overload is documented using a separate py:function directive since they each have their own documentation.

I'm putting the docstrings in the binding code so users can still use the REPL's `help()` function. pybind11 adds a basic signature to the exposed members even when docstr isn't specified in the binding; I turned this pybind11 feature is turned off for various reasons. It wouldn't be hard to change where the docstr is found, but I don't want to maintain more than 1 docs src.

It's a shame that autodoc doesn't support stub files yet.

**Comment by jbms** [comment]
To be clear, with tensorstore all of the documentation is in the docstring.  If you have multiple overloads, pybind11 combines all of the signatures and their docstrings into a single docstring for the real Python function that dispatches to the various overloads.  The tensorstore autosummary extension parses the pybind11 docstring format and extracts the individual overloads., and then in a hacky way uses autodoc to document each of them as separate functions.

https://github.com/google/tensorstore/blob/fe7aae44a788dbdfc731ea1c00b83f0d51e4f2f4/python/tensorstore/context.cc#L94

**Comment by 2bndy5** [comment]
```
Overload:
  spec
```
I'm guessing this is your overload id.

> in a hacky way uses autodoc

I'm guessing that you didn't "settle" for the `:no-index:` option - that would be too easy 😆 .

**Comment by jbms** [comment]
Actually the tensorstore autosummary does always use the noindex option, but then adds the ids and registers the object with the Python domain itself.

**Comment by 2bndy5** [comment]
@jbms Using the fix for this (included in v0.6.1) I can now document overloaded function with pybind11 like so:
### C++ src (binding code)
```c++
PYBIND11_MODULE(rf24_network, m)
{
    m.doc() = "A Python module that wraps the RF24Network C++ library's API";
    py::options options;
    options.disable_function_signatures();

    py::class_<RF24NetworkHeader>(m, "RF24NetworkHeader")

       // ...

        .def("peek", &RF24NetworkWrapper::peek_header, R"docstr(
            peek(header: RF24NetworkHeader) -> int \
            peek(maxlen: int) -> Tuple[RF24NetworkHeader, bytearray]
            To fetch the next available frame's header received by the network node, the parameter and return type is as follows:

            :param RF24NetworkHeader header: The object to save the header information to.
            :Returns: The frame's message size (not including the 8-byte header saved to the ``header`` parameter).
        )docstr",
             py::arg("header"))

        // *****************************************************************************

        .def("peek", &RF24NetworkWrapper::peek_frame, R"docstr(
            To fetch the next available frame received by the network node, the parameter and return type is as follows:

            :param int maxlen: The maximum length of the message to fetch. If this parameter is unspecified or greater than
                the actual frame's message size, then only the frame's full message size is used.

            :Returns: A 2-tuple containing the frame's header (of type `RF24NetworkHeader`) and the frame's message (of type `bytearray`).
        )docstr",
             py::arg("maxlen") = MAX_PAYLOAD_SIZE)

       // ...
      ;
}
```
The important part here is the trailing LF escape (`\`) in the multi-lined signature. Also notice the `options.disable_function_signatures();` used to add proper python type hinting.

### then using autodoc
```rst
.. autoclass:: pyrf24.rf24_network.RF24Network
    :members: begin, update, available, peek, read, write, node_address
    :special-members: __init__
```

With the above approach and v0.6.0, I can reproduce this issue.
